{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e91a6d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "097c2d8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = \"final.fasta\"  # Replace with the path to your FASTA file\n",
    "records = SeqIO.parse(fasta_file, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9f1f248c",
   "metadata": {},
   "outputs": [],
   "source": [
    "header_lines = []\n",
    "sequences = []\n",
    "with open(\"final.fasta\", \"r\") as fasta_file:\n",
    "    header = None\n",
    "    sequence = \"\"\n",
    "    for line in fasta_file:\n",
    "        if line.startswith(\">\"):\n",
    "            # This is a header line\n",
    "            if header is not None:\n",
    "                header_lines.append(header)\n",
    "                sequences.append(sequence)\n",
    "            header = line.strip()\n",
    "            sequence = \"\"\n",
    "        else:\n",
    "            # This is sequence data\n",
    "            sequence += line.strip()\n",
    "\n",
    "    # Append the last record\n",
    "    if header is not None:\n",
    "        header_lines.append(header)\n",
    "        sequences.append(sequence)\n",
    "descriptions=[]\n",
    "with open('final.fasta', 'r') as file:\n",
    "    for line in file:\n",
    "        if line.startswith('>'):\n",
    "            # Extract the description part\n",
    "            if line.startswith('>'):\n",
    "                # Split the line by '|', and then further split by 'OS' to extract the description\n",
    "                parts = line.split('|')\n",
    "                if len(parts) >= 3:\n",
    "                    description_up_to_OS = parts[2].split(' OS')[0].strip()\n",
    "                    descriptions.append(description_up_to_OS)\n",
    "# Create a DataFrame\n",
    "data = {\"Header Line\": header_lines, \"Sequence\": sequences, 'Descriptions': descriptions}\n",
    "df = pd.DataFrame(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2e23b143",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Full Header:\n",
      ">PA2A6_CROHD | Crotalus | Acidic phospholipase A2 CH-E6' OS=Crotalus horridus OX=35024 PE=1 SV=1\n"
     ]
    }
   ],
   "source": [
    "# Assuming you have a DataFrame named df\n",
    "\n",
    "# Access the full header (entire header line) of the first entry in the DataFrame\n",
    "full_header = df.loc[0, \"Header Line\"]\n",
    "\n",
    "# Print the full header\n",
    "print(\"Full Header:\")\n",
    "print(full_header)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "fe971ac3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Header Line</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Descriptions</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;PA2A6_CROHD | Crotalus | Acidic phospholipase...</td>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 CH-E6'</td>\n",
       "      <td>crotalus horridus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;PA2A_CROAD | Crotalus | Acidic phospholipase ...</td>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 beta</td>\n",
       "      <td>crotalus adamanteus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;PA2A_CROAT | Crotalus | Acidic phospholipase ...</td>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2</td>\n",
       "      <td>crotalus atrox</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;PA2AA_CROVV | Crotalus | Acidic phospholipase...</td>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 Cvv-E6a</td>\n",
       "      <td>crotalus viridis viridis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;PA2AD_CROOA | Crotalus | Acidic phospholipase...</td>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 CoaPLA2</td>\n",
       "      <td>crotalus oreganus abyssus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>&gt;A0A6P5JCK6_PHACI | Phascolarctos | Phospholip...</td>\n",
       "      <td>VEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRC...</td>\n",
       "      <td>Phospholipase A2</td>\n",
       "      <td>phascolarctos cinereus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>&gt;A0A1S3GT82_DIPOR | Dipodomys | Phospholipase ...</td>\n",
       "      <td>MRTLWIVAVLL----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYGC...</td>\n",
       "      <td>Phospholipase A2</td>\n",
       "      <td>dipodomys ordii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>&gt;PA2HB_BOTLC | Bothrops | Basic phospholipase ...</td>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Basic phospholipase A2 homolog blK-PLA2</td>\n",
       "      <td>bothrops leucurus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>&gt;PA2A2_BITNA | Bitis | Acidic phospholipase A2...</td>\n",
       "      <td>LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFVH...</td>\n",
       "      <td>Acidic phospholipase A2 CM-II</td>\n",
       "      <td>bitis nasicornis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>&gt;G1SRA0_RABIT | Oryctolagus | Phospholipase A2...</td>\n",
       "      <td>MRTLWIVAVL----LLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGC...</td>\n",
       "      <td>Phospholipase A2</td>\n",
       "      <td>oryctolagus cuniculus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Header Line  \\\n",
       "0    >PA2A6_CROHD | Crotalus | Acidic phospholipase...   \n",
       "1    >PA2A_CROAD | Crotalus | Acidic phospholipase ...   \n",
       "2    >PA2A_CROAT | Crotalus | Acidic phospholipase ...   \n",
       "3    >PA2AA_CROVV | Crotalus | Acidic phospholipase...   \n",
       "4    >PA2AD_CROOA | Crotalus | Acidic phospholipase...   \n",
       "..                                                 ...   \n",
       "246  >A0A6P5JCK6_PHACI | Phascolarctos | Phospholip...   \n",
       "247  >A0A1S3GT82_DIPOR | Dipodomys | Phospholipase ...   \n",
       "248  >PA2HB_BOTLC | Bothrops | Basic phospholipase ...   \n",
       "249  >PA2A2_BITNA | Bitis | Acidic phospholipase A2...   \n",
       "250  >G1SRA0_RABIT | Oryctolagus | Phospholipase A2...   \n",
       "\n",
       "                                              Sequence  \\\n",
       "0    MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "1    MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "2    MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "3    MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "4    MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "..                                                 ...   \n",
       "246  VEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRC...   \n",
       "247  MRTLWIVAVLL----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYGC...   \n",
       "248  SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "249  LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFVH...   \n",
       "250  MRTLWIVAVL----LLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGC...   \n",
       "\n",
       "                                Descriptions                    Species  \n",
       "0             Acidic phospholipase A2 CH-E6'          crotalus horridus  \n",
       "1               Acidic phospholipase A2 beta        crotalus adamanteus  \n",
       "2                    Acidic phospholipase A2             crotalus atrox  \n",
       "3            Acidic phospholipase A2 Cvv-E6a   crotalus viridis viridis  \n",
       "4            Acidic phospholipase A2 CoaPLA2  crotalus oreganus abyssus  \n",
       "..                                       ...                        ...  \n",
       "246                         Phospholipase A2     phascolarctos cinereus  \n",
       "247                         Phospholipase A2            dipodomys ordii  \n",
       "248  Basic phospholipase A2 homolog blK-PLA2          bothrops leucurus  \n",
       "249            Acidic phospholipase A2 CM-II           bitis nasicornis  \n",
       "250                         Phospholipase A2      oryctolagus cuniculus  \n",
       "\n",
       "[251 rows x 4 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a new column \"OS\" in the DataFrame by extracting the OS information until \"OX\" from the \"Header Line\"\n",
    "df['Species'] = df['Header Line'].str.extract(r'OS=(.*?)(?= OX)')\n",
    "df['Species']=df['Species'].str.lower()\n",
    "# Print the DataFrame to see the new \"OS\" column\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0ece5cf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "sequences=[Seq(seq) for seq in df['Sequence']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "cdc16f8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align.Applications import ClustalOmegaCommandline\n",
    "\n",
    "# Create a temporary FASTA file containing the sequences\n",
    "with open(\"input_sequences.fasta\", \"w\") as fasta_file:\n",
    "    for i, seq in enumerate(sequences):\n",
    "        fasta_file.write(f\">Seq{i}\\n{seq}\\n\")\n",
    "\n",
    "# Define the output file path\n",
    "output_file = \"output_alignment.fasta\"\n",
    "\n",
    "# Create a command-line instance for Clustal Omega\n",
    "clustalomega_cline = ClustalOmegaCommandline(infile=\"input_sequences.fasta\", outfile=output_file, outfmt=\"fasta\")\n",
    "\n",
    "# Run Clustal Omega\n",
    "stdout, stderr = clustalomega_cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "14c81eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences = {}\n",
    "current_seq = \"\"\n",
    "\n",
    "    # Open and read the FASTA file\n",
    "with open('output_alignment.fasta', 'r') as file:\n",
    "    for line in file:\n",
    "        line = line.strip()\n",
    "            # Check if the line is a header line\n",
    "        if line.startswith('>'):\n",
    "            if current_seq:\n",
    "                sequences[header] = current_seq\n",
    "                current_seq = \"\"\n",
    "            header = line[1:]\n",
    "        else:\n",
    "            current_seq += line\n",
    "        # Add the last sequence\n",
    "    if current_seq:\n",
    "        sequences[header] = current_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0228041d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Header Line</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Descriptions</th>\n",
       "      <th>Species</th>\n",
       "      <th>Aligned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;PA2A6_CROHD | Crotalus | Acidic phospholipase...</td>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 CH-E6'</td>\n",
       "      <td>crotalus horridus</td>\n",
       "      <td>MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;PA2A_CROAD | Crotalus | Acidic phospholipase ...</td>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 beta</td>\n",
       "      <td>crotalus adamanteus</td>\n",
       "      <td>MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;PA2A_CROAT | Crotalus | Acidic phospholipase ...</td>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2</td>\n",
       "      <td>crotalus atrox</td>\n",
       "      <td>MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;PA2AA_CROVV | Crotalus | Acidic phospholipase...</td>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 Cvv-E6a</td>\n",
       "      <td>crotalus viridis viridis</td>\n",
       "      <td>MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;PA2AD_CROOA | Crotalus | Acidic phospholipase...</td>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 CoaPLA2</td>\n",
       "      <td>crotalus oreganus abyssus</td>\n",
       "      <td>MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Header Line  \\\n",
       "0  >PA2A6_CROHD | Crotalus | Acidic phospholipase...   \n",
       "1  >PA2A_CROAD | Crotalus | Acidic phospholipase ...   \n",
       "2  >PA2A_CROAT | Crotalus | Acidic phospholipase ...   \n",
       "3  >PA2AA_CROVV | Crotalus | Acidic phospholipase...   \n",
       "4  >PA2AD_CROOA | Crotalus | Acidic phospholipase...   \n",
       "\n",
       "                                            Sequence  \\\n",
       "0  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "1  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "2  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "3  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "4  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "\n",
       "                      Descriptions                    Species  \\\n",
       "0   Acidic phospholipase A2 CH-E6'          crotalus horridus   \n",
       "1     Acidic phospholipase A2 beta        crotalus adamanteus   \n",
       "2          Acidic phospholipase A2             crotalus atrox   \n",
       "3  Acidic phospholipase A2 Cvv-E6a   crotalus viridis viridis   \n",
       "4  Acidic phospholipase A2 CoaPLA2  crotalus oreganus abyssus   \n",
       "\n",
       "                                             Aligned  \n",
       "0  MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "1  MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "2  MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "3  MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "4  MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.DataFrame(list(sequences.items()), columns=['Header', 'Sequence'])\n",
    "df['Aligned']=df1['Sequence']\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "063063d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop_duplicates(subset='Sequence', keep='first',inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "91e3f361",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Descriptions</th>\n",
       "      <th>Species</th>\n",
       "      <th>Aligned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 CH-E6'</td>\n",
       "      <td>crotalus horridus</td>\n",
       "      <td>MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 Tgc-E6</td>\n",
       "      <td>trimeresurus gracilis</td>\n",
       "      <td>MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 S1E6-c</td>\n",
       "      <td>calloselasma rhodostoma</td>\n",
       "      <td>MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Acidic phospholipase A2</td>\n",
       "      <td>deinagkistrodon acutus</td>\n",
       "      <td>---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 DE-I</td>\n",
       "      <td>ovophis okinavensis</td>\n",
       "      <td>MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Sequence  \\\n",
       "0   MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "11  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "13  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "24  SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "26  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "\n",
       "                      Descriptions                  Species  \\\n",
       "0   Acidic phospholipase A2 CH-E6'        crotalus horridus   \n",
       "11  Acidic phospholipase A2 Tgc-E6    trimeresurus gracilis   \n",
       "13  Acidic phospholipase A2 S1E6-c  calloselasma rhodostoma   \n",
       "24         Acidic phospholipase A2   deinagkistrodon acutus   \n",
       "26    Acidic phospholipase A2 DE-I      ovophis okinavensis   \n",
       "\n",
       "                                              Aligned  \n",
       "0   MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "11  MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "13  MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "24  ---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "26  MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=df.drop('Header Line',axis=1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ccbbe151",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop_duplicates(subset=\"Species\", keep='first', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "08f130ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Descriptions</th>\n",
       "      <th>Species</th>\n",
       "      <th>Aligned</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 CH-E6'</td>\n",
       "      <td>crotalus horridus</td>\n",
       "      <td>MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 Tgc-E6</td>\n",
       "      <td>trimeresurus gracilis</td>\n",
       "      <td>MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 S1E6-c</td>\n",
       "      <td>calloselasma rhodostoma</td>\n",
       "      <td>MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Acidic phospholipase A2</td>\n",
       "      <td>deinagkistrodon acutus</td>\n",
       "      <td>---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 DE-I</td>\n",
       "      <td>ovophis okinavensis</td>\n",
       "      <td>MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Acidic phospholipase A2 A'</td>\n",
       "      <td>gloydius halys</td>\n",
       "      <td>---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Acidic phospholipase A2</td>\n",
       "      <td>gloydius ussuriensis</td>\n",
       "      <td>---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFVH...</td>\n",
       "      <td>Acidic phospholipase A2 4</td>\n",
       "      <td>trimeresurus gramineus</td>\n",
       "      <td>----------------------LVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Basic phospholipase A2 APP-D49</td>\n",
       "      <td>agkistrodon piscivorus piscivorus</td>\n",
       "      <td>---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Neutral phospholipase A2 ammodytin I2</td>\n",
       "      <td>vipera ammodytes ammodytes</td>\n",
       "      <td>MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Basic phospholipase A2 Drk-b2</td>\n",
       "      <td>daboia russelii</td>\n",
       "      <td>MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Acidic phospholipase A2 2</td>\n",
       "      <td>bothrops asper</td>\n",
       "      <td>---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 BpPLA2-TXI</td>\n",
       "      <td>bothrops pauloensis</td>\n",
       "      <td>MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFVH...</td>\n",
       "      <td>Acidic phospholipase A2</td>\n",
       "      <td>bothrops jararaca</td>\n",
       "      <td>----------------------LVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Basic phospholipase A2 DsM-S1</td>\n",
       "      <td>daboia siamensis</td>\n",
       "      <td>MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Acidic phospholipase A2 5</td>\n",
       "      <td>echis pyramidum leakeyi</td>\n",
       "      <td>MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>MRTLWIVAVLL----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYGC...</td>\n",
       "      <td>Phospholipase A2</td>\n",
       "      <td>pogona vitticeps</td>\n",
       "      <td>MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFVH...</td>\n",
       "      <td>Acidic phospholipase A2 PhTX-III (Fragments)</td>\n",
       "      <td>bothrocophias hyoprora</td>\n",
       "      <td>----------------------LVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Acidic phospholipase A2 PLA-1</td>\n",
       "      <td>eristicophis macmahoni</td>\n",
       "      <td>---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Basic phospholipase A2 (Fragments)</td>\n",
       "      <td>agkistrodon piscivorus leucostoma</td>\n",
       "      <td>---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Phospholipase A2 homolog 1</td>\n",
       "      <td>bothrops atrox</td>\n",
       "      <td>MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Acidic phospholipase A2 braziliase-II (Fragments)</td>\n",
       "      <td>bothrops brazili</td>\n",
       "      <td>---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...</td>\n",
       "      <td>Basic phospholipase A2 BP-III</td>\n",
       "      <td>protobothrops flavoviridis</td>\n",
       "      <td>MRTLWIVAVLLL-----GVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...</td>\n",
       "      <td>Basic phospholipase A2 9</td>\n",
       "      <td>crotalus durissus cumanensis</td>\n",
       "      <td>---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>VLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQD...</td>\n",
       "      <td>phospholipase A2</td>\n",
       "      <td>naja naja</td>\n",
       "      <td>--------VLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>MRTLWIVAVLLLGV---EGSLVQFEMMIMEVAKRSGLLWYSAYGCY...</td>\n",
       "      <td>Phospholipase A2 group IIA</td>\n",
       "      <td>pseudonaja textilis</td>\n",
       "      <td>MRTLWIVAVLLLGV-----EGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>GSLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCF...</td>\n",
       "      <td>Basic phospholipase A2 VRV-PL-VIIIa-like</td>\n",
       "      <td>notechis scutatus</td>\n",
       "      <td>--------------------GSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHG-RPQDATDRCCFV...</td>\n",
       "      <td>Basic phospholipase A2 BmjeTX-I</td>\n",
       "      <td>bothrops marajoensis</td>\n",
       "      <td>----------------------LVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>GSLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCF...</td>\n",
       "      <td>Phosphatidylcholine 2-acylhydrolase</td>\n",
       "      <td>laticauda laticaudata</td>\n",
       "      <td>--------------------GSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFVH...</td>\n",
       "      <td>Acidic phospholipase A2 CM-II</td>\n",
       "      <td>bitis nasicornis</td>\n",
       "      <td>----------------------LVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>MRTLWIVAVL----LLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGC...</td>\n",
       "      <td>Phospholipase A2</td>\n",
       "      <td>oryctolagus cuniculus</td>\n",
       "      <td>MRTLWIVAVL----L-LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Sequence  \\\n",
       "0    MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "11   MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "13   MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "24   SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "26   MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "35   SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "40   SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "44   LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFVH...   \n",
       "46   SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "55   MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "59   MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "76   SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "78   MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "90   LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFVH...   \n",
       "100  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "105  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "131  MRTLWIVAVLL----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYGC...   \n",
       "137  LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFVH...   \n",
       "138  SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "141  SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "143  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "148  SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "156  MRTLWIVAVLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGW...   \n",
       "181  SLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFV...   \n",
       "191  VLLLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQD...   \n",
       "194  MRTLWIVAVLLLGV---EGSLVQFEMMIMEVAKRSGLLWYSAYGCY...   \n",
       "195  GSLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCF...   \n",
       "222  LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHG-RPQDATDRCCFV...   \n",
       "243  GSLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCF...   \n",
       "249  LVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGHGRPQDATDRCCFVH...   \n",
       "250  MRTLWIVAVL----LLGVEGSLVQFEMMIMEVAKRSGLLWYSAYGC...   \n",
       "\n",
       "                                          Descriptions  \\\n",
       "0                       Acidic phospholipase A2 CH-E6'   \n",
       "11                      Acidic phospholipase A2 Tgc-E6   \n",
       "13                      Acidic phospholipase A2 S1E6-c   \n",
       "24                             Acidic phospholipase A2   \n",
       "26                        Acidic phospholipase A2 DE-I   \n",
       "35                          Acidic phospholipase A2 A'   \n",
       "40                             Acidic phospholipase A2   \n",
       "44                           Acidic phospholipase A2 4   \n",
       "46                      Basic phospholipase A2 APP-D49   \n",
       "55               Neutral phospholipase A2 ammodytin I2   \n",
       "59                       Basic phospholipase A2 Drk-b2   \n",
       "76                           Acidic phospholipase A2 2   \n",
       "78                  Acidic phospholipase A2 BpPLA2-TXI   \n",
       "90                             Acidic phospholipase A2   \n",
       "100                      Basic phospholipase A2 DsM-S1   \n",
       "105                          Acidic phospholipase A2 5   \n",
       "131                                   Phospholipase A2   \n",
       "137       Acidic phospholipase A2 PhTX-III (Fragments)   \n",
       "138                      Acidic phospholipase A2 PLA-1   \n",
       "141                 Basic phospholipase A2 (Fragments)   \n",
       "143                         Phospholipase A2 homolog 1   \n",
       "148  Acidic phospholipase A2 braziliase-II (Fragments)   \n",
       "156                      Basic phospholipase A2 BP-III   \n",
       "181                           Basic phospholipase A2 9   \n",
       "191                                   phospholipase A2   \n",
       "194                         Phospholipase A2 group IIA   \n",
       "195           Basic phospholipase A2 VRV-PL-VIIIa-like   \n",
       "222                    Basic phospholipase A2 BmjeTX-I   \n",
       "243                Phosphatidylcholine 2-acylhydrolase   \n",
       "249                      Acidic phospholipase A2 CM-II   \n",
       "250                                   Phospholipase A2   \n",
       "\n",
       "                               Species  \\\n",
       "0                    crotalus horridus   \n",
       "11               trimeresurus gracilis   \n",
       "13             calloselasma rhodostoma   \n",
       "24              deinagkistrodon acutus   \n",
       "26                 ovophis okinavensis   \n",
       "35                      gloydius halys   \n",
       "40                gloydius ussuriensis   \n",
       "44              trimeresurus gramineus   \n",
       "46   agkistrodon piscivorus piscivorus   \n",
       "55          vipera ammodytes ammodytes   \n",
       "59                     daboia russelii   \n",
       "76                      bothrops asper   \n",
       "78                 bothrops pauloensis   \n",
       "90                   bothrops jararaca   \n",
       "100                   daboia siamensis   \n",
       "105            echis pyramidum leakeyi   \n",
       "131                   pogona vitticeps   \n",
       "137             bothrocophias hyoprora   \n",
       "138             eristicophis macmahoni   \n",
       "141  agkistrodon piscivorus leucostoma   \n",
       "143                     bothrops atrox   \n",
       "148                   bothrops brazili   \n",
       "156         protobothrops flavoviridis   \n",
       "181       crotalus durissus cumanensis   \n",
       "191                          naja naja   \n",
       "194                pseudonaja textilis   \n",
       "195                  notechis scutatus   \n",
       "222               bothrops marajoensis   \n",
       "243              laticauda laticaudata   \n",
       "249                   bitis nasicornis   \n",
       "250              oryctolagus cuniculus   \n",
       "\n",
       "                                               Aligned  \n",
       "0    MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "11   MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "13   MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "24   ---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "26   MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "35   ---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "40   ---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "44   ----------------------LVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "46   ---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "55   MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "59   MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "76   ---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "78   MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "90   ----------------------LVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "100  MRTLWIVAVLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "105  MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "131  MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "137  ----------------------LVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "138  ---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "141  ---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "143  MRTLWIVAVLL-----LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "148  ---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "156  MRTLWIVAVLLL-----GVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "181  ---------------------SLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "191  --------VLLLG-----VEGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "194  MRTLWIVAVLLLGV-----EGSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "195  --------------------GSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "222  ----------------------LVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "243  --------------------GSLVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "249  ----------------------LVQFEMMIMEVAKRSGLLWYSAYG...  \n",
       "250  MRTLWIVAVL----L-LGVEGSLVQFEMMIMEVAKRSGLLWYSAYG...  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexes_to_exclude=[209,216,226,228,238,245,246,247]\n",
    "df_filtered = df.drop(indexes_to_exclude)\n",
    "df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "be03030c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nexus file 'aligned_sequences.nex' has been created.\n"
     ]
    }
   ],
   "source": [
    "fasta_file = \"aligned_sequences.fasta\"\n",
    "\n",
    "with open(fasta_file, \"w\") as handle:\n",
    "    for idx, row in df_filtered.iterrows():\n",
    "        sequence = row[\"Aligned\"]\n",
    "        species = row[\"Species\"]\n",
    "        record = SeqIO.SeqRecord(Seq(sequence), id=species, description=\"\")\n",
    "        SeqIO.write(record, handle, \"fasta\")\n",
    "\n",
    "# Step 2: Create a Nexus file using the generated FASTA file\n",
    "nexus_file = \"aligned_sequences.nex\"\n",
    "nexus_data = \"\"\"#NEXUS\n",
    "BEGIN DATA;\n",
    "DIMENSIONS NTAX={ntax} NCHAR={nchar};\n",
    "FORMAT DATATYPE=protein MISSING=? GAP=-;\n",
    "MATRIX\n",
    "{matrix}\n",
    ";\n",
    "END;\n",
    "\"\"\"\n",
    "\n",
    "ntax = len(df_filtered)\n",
    "nchar = len(df_filtered[\"Aligned\"][0])  # Assuming all sequences have the same length\n",
    "matrix = []\n",
    "\n",
    "with open(fasta_file, \"r\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        header = record.description\n",
    "        sequence = str(record.seq)\n",
    "        matrix.append(f\"{header} {sequence}\")\n",
    "nexus_data = nexus_data.format(ntax=ntax, nchar=nchar, matrix=\"\\n\".join(matrix))\n",
    "\n",
    "with open(nexus_file, \"w\") as handle:\n",
    "    handle.write(nexus_data)\n",
    "\n",
    "print(f\"Nexus file '{nexus_file}' has been created.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "13d735c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('MRTLWIVAVL----L-LGVEGSLVQFEMMIMEVAKRSGLLWYSAYGCYCGWGGH...EPC'), id='oryctolagus', name='oryctolagus', description='oryctolagus cuniculus', dbxrefs=[])"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
